### PR TITLE
8285515: (dc) DatagramChannel.disconnect fails with "Invalid argument" on macOS 12.4 beta2

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change with proposes to address the issue reported in https://bugs.openjdk.java.net/browse/JDK-8285515?

We have noticed that in the latest 12.4 Beta2 version of MacOSX, the implementation of `DatagramChannelImpl.disconnect` fails with an unexpected error. 

Internally, the disconnect is implemented as a call to `connect` by passing it a "null" address as suggested in the documentation of `man connect`. In previous versions of MacOSX, this call used to return a `EADDRNOTAVAIL` errno and at the same time would correctly dissolve the connected association. The `EADDRNOTAVAIL` was then caught and intentionally ignored by the code in the JNI layer in `DatagramChannelImpl.disconnect0`.

In MacOSX 12.4 Beta 2 we are now seeing this `connect` call return `EINVAL`. This change in behaviour is noticed only for IPv4-mapped IPv6 addresses.

The `man connect` documentation on MacOSX states that even `disconnectx` can be used to disconnect datagram sockets. The commit here does that change to use `disconnectx` for MacOSX. Tests have been successfully run on some older Mac setups  to make sure this change passes there as well.

Currently `tier1`, `tier2` and `tier3` run is in progress across various OS.

No additional tests have been added for this change since the many existing (failing) tests already cover this use case.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8285515](https://bugs.openjdk.java.net/browse/JDK-8285515): (dc) DatagramChannel.disconnect fails with "Invalid argument" on macOS 12.4 beta2

### Reviewers
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8445/head:pull/8445` \
`$ git checkout pull/8445`

Update a local copy of the PR: \
`$ git checkout pull/8445` \
`$ git pull https://git.openjdk.java.net/jdk pull/8445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8445`

View PR using the GUI difftool: \
`$ git pr show -t 8445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8445.diff">https://git.openjdk.java.net/jdk/pull/8445.diff</a>

</details>

### CODE REVIEW

File: src/java.base/unix/native/libnio/ch/DatagramChannelImpl.c [hunk 1]
```diff
@@ -50,20 +50,28 @@ Java_sun_nio_ch_DatagramChannelImpl_disconnect0(JNIEnv *env, jclass clazz,
     jint fd = fdval(env, fdo);
     int rv;
 
+#if defined(__APPLE__)
+    // On MacOSX systems we use disconnectx
```
Comment (AlanBateman) [inline src/java.base/unix/native/libnio/ch/DatagramChannelImpl.c:54]: Thanks for doing this. Do you mind changing this comment to "macOS" as I think that is how the OS is named these days.

Comment (jaikiran) [reply src/java.base/unix/native/libnio/ch/DatagramChannelImpl.c]: Done. I have pushed an update to this PR with this change.

**Comment by dfuch** [comment]
That looks good to me. I'd suggest running some test repeat campaign on the CI to make sure this works as expected on all macOS version that we support, and there's no regression on other OS. 
I would also suggest tagging the tests that failed without this change with `@bug 8285515` (even though they only failed when running on the 12.4 beta2)

**Comment by AlanBateman** [comment]
> I would also suggest tagging the tests that failed without this change with `@bug 8285515` (even though they only failed when running on the 12.4 beta2)

Maybe okay to add it to one test for disconnect but please don't change every test that failed, it's just not helpful.

**Comment by jaikiran** [comment]
> > I would also suggest tagging the tests that failed without this change with `@bug 8285515` (even though they only failed when running on the 12.4 beta2)
> 
> Maybe okay to add it to one test for disconnect but please don't change every test that failed, it's just not helpful.

Done. I updated the PR to add it to just one test case.

**Comment by jaikiran** [comment]
> I'd suggest running some test repeat campaign on the CI to make sure this works as expected on all macOS versions that we support, and there's no regression on other OS. 

tier1, tier2, tier3 tests that were in progress against various different OS, completed without any related failures.

I've additionally triggered a run with test repeat of 50, solely against macosx-x64 and macosx-aarch64. I'll let them run overnight and see how they go.

**Review by dfuch — APPROVED** [review]
LGTM

**Review by AlanBateman — APPROVED** [review]

**Comment by jaikiran** [comment]
The test runs completed without any related failures. Thank you Alan and Daniel for the reviews.

**Pull request closed without merging**